Artifacts/windows-vsts-build-agent/vsts-agent-install.ps1

```diff
@@ -1,148 +1,293 @@
+#       This is necessary to ensure we capture errors inside the try-catch-finally block.
 $ErrorActionPreference = "Stop"
 
-trap
+# Ensure we set the working directory to that of the script.
+pushd $PSScriptRoot
+
+# Configure strict debugging.
+Set-PSDebug -Strict
+
+###################################################################################################
+
+#
+# Functions used in this script.
+#
+
+function Handle-LastError
 {
-    $_ | Write-Error -ErrorAction Continue
-    exit 1
+    [CmdletBinding()]
+    param(
+    )
+
+    $message = $error[0].Exception.Message
+    if ($message)
+    {
+        Write-Host -Object "ERROR: $message" -ForegroundColor Red
+    }
+    
+    # IMPORTANT NOTE: Throwing a terminating error (using $ErrorActionPreference = "Stop") still
+    # returns exit code zero from the PowerShell script when using -File. The workaround is to
+    # NOT use -File when calling this script and leverage the try-catch-finally block and return
+    # a non-zero exit code from the catch block.
+    exit -1
 }
 
-if ($vstsAccount -match "https*://" -or $vstsAccount -match "visualstudio.com")
+function Test-Parameters
 {
-    Write-Error "VSTS account should not be the URL, just the account name."
+    [CmdletBinding()]
+    param(
+        [string] $VstsAccount,
+        [string] $WorkDirectory
+    )
+    
+    if ($VstsAccount -match "https*://" -or $VstsAccount -match "visualstudio.com")
+    {
+        Write-Error "VSTS account '$VstsAccount' should not be the URL, just the account name."
+    }
+
+    if (![string]::IsNullOrWhiteSpace($WorkDirectory) -and !(Test-ValidPath -Path $WorkDirectory))
+    {
+        Write-Error "Work directory '$WorkDirectory' is not a valid path."
+    }
 }
 
-if ($workDirectory -ne "" -and !(Test-Path -Path $workDirectory -IsValid -ErrorAction Ignore))
+function Test-ValidPath
 {
-    Write-Error "Work Directory '$workDirectory' is not a valid path."
+    param(
+        [string] $Path
+    )
+    
+    $isValid = Test-Path -Path $Path -IsValid -PathType Container
+    
+    try
+    {
+        [IO.Path]::GetFullPath($workDirectory) | Out-Null
+    }
+    catch
+    {
+        $isValid = $false
+    }
+    
+    return $isValid
 }
 
-$currentLocation = Split-Path -parent $MyInvocation.MyCommand.Definition
+function Download-AgentPackage
+{
+    [CmdletBinding()]
+    param(
+        [string] $VstsAccount,
+        [string] $VstsUserPassword
+    )
+    
+    # Create a temporary directory to download from VSTS the agent package (agent.zip) to, and then launch the configuration.
+    $agentTempFolderName = Join-Path $env:temp ([System.IO.Path]::GetRandomFileName())
+    New-Item -ItemType Directory -Force -Path $agentTempFolderName | Out-Null
+
+    $agentPackagePath = "$agentTempFolderName\agent.zip"
+    $serverUrl = "https://$VstsAccount.visualstudio.com"
+    $vstsAgentUrl = $serverUrl + '/_apis/distributedtask/packages/agent/win7-x64?$top=1&api-version=3.0'

```

Comment:
Why use concatination here? Couldn't you use string interpolation as line 109 and 110?